Feature aggregate consumption

lomamech opened the pull request:
````
You should reload database and run the yarn migrate command
````
closes #4850

lomamech committed:
feature(Record aggregate consumption)
- Implement client side for record aggregate consumption

closes #4850

lomamech committed:
Implement aggregate consumption

lomamech committed:
Implement API for aggregate consumption

lomamech committed:
Stock: implement and improve aggregate consumption

lomamech committed:
Improved management of stock-outs in aggregate stock consumption

lomamech committed:
feature(Stock Exit Report)
- Add aggregation consumption option in Stock exit report

lomamech committed:
Throw invalid data

lomamech committed:
test(Aggregate consumption)
- Implement integration test for Aggregate consumption

lomamech committed:
Implement E2E test

lomamech committed:
test(Aggregate consumption)
- Implement test E2E for aggregate consumption

lomamech committed:
Add script in migrate file

lomamech committed:
Reset value for E2E of stock

lomamech committed:
Test: reset value for e2e test of stock

lomamech committed:
Fix and resolve conflict

lomamech committed:
Clean and sanitaze code

lomamech committed:
Improve french grammar

jmcameron reviewed (commented):
I did a crude test (with the IMCK db).  But when I clicked on the [Submit] button, nothing happened and I see this error:

`
vendor.min.js:15710 TypeError: Cannot read property 'length' of undefined
    at getCheckedNodes (bhima.min.js:31917)
    at bhCheckboxTreeController.$ctrl.$onChanges (bhima.min.js:31839)
`

This is on line 132 of bhCheckboxTree.js

Also noticed a problem in the migration file: It looks like the migration file defines the unit entry for the aggregated page twice.

lomamech committed:
test(Unit test)
- fix unit test for the validate will block when total quantity is over
  max quantity

lomamech committed:
Sanitaze migrate file

jniles commented on client/src/i18n/fr/report.json:
```diff
@@ -244,7 +244,9 @@
-      "INCLUDE_INFORMATION" : "Information à inclure",
+      "INCLUDE_AGGREGATE_CONSUMPTION": "Inclure les consommations agrégées",
+	  "INCLUDE_INFORMATION" : "Information à inclure",
+	  "INCLUDE_INFORMATION" : "Information à inclure",
```
Looks like this is repeated twice.  Could you remove the duplicate?

jniles commented on client/src/i18n/fr/stock.json:
```diff
@@ -221,6 +237,7 @@
     "ENABLE_STRICT_DEPOT_DISTRIBUTION_HELP_TEXT" : "Limite les dépôts qui peuvent recevoir du stock à partir d'un dépôt donné; lors de la sortie de stock vers un dépôt, seul les dépôts autorisés peuvent s'afficher pour le dépôt donné afin de recevoir du stock"
   },
   "STOCK_FLUX" : {
+    "AGGREGATE_CONSUMPTION" : "Consommation agrégée",
```
I agree with the decision to make a new flux_type.  Note that we'll need to update all our procedures that create CMM to say this is a "consumption" type, no matter the depot type.

jniles commented on server/controllers/stock/index.js:
```diff
@@ -975,3 +977,221 @@ function getStockTransfers(req, res, next) {
+    }
+
+    const checkInvalid = movement.lots
+      .filter(l => ((l.quantity_consumed + l.quantity_lost) > l.oldQuantity));
```
I think this check should be `>=`, not `>`.

lomamech replied:
Here I check the invalid situation, and this case is very unlikely because on the client side prevents the validation of an invalid form

jniles replied:
Oh!  You are right.  I misread this.  Thanks for the clarification. 👍

jniles commented on server/controllers/stock/index.js:
```diff
@@ -975,3 +977,221 @@ function getStockTransfers(req, res, next) {
+      .filter(l => ((l.quantity_consumed + l.quantity_lost) > l.oldQuantity));
+
+    if (checkInvalid.length) {
+      throw new Error('Invalid data');
```
We could improve this error message:
```js
throw new Error('Invalid data!  Some lots have consumed or lost more stock than they originally had.'); 
```

jniles commented on client/src/modules/stock/aggregated_consumption/modals/consumption_lot.modal.html:
```diff
@@ -0,0 +1,89 @@
+            type="checkbox"
+            ng-true-value="1"
+            ng-false-value="0"> 
+          <span translate>LOTS.ENABLE_FAST_INSERT</span>
```
This checkbox doesn't make sense here.  We aren't dealing with lots.  We either need to remove it, or change the text and potentially behavior to reflect what we are really doing.

jniles commented on client/src/modules/stock/aggregated_consumption/templates/days_stock_out.tmpl.html:
```diff
@@ -0,0 +1,12 @@
+  <input
+    ng-if="row.groupHeader"
+    name="days_stock_out"
+    id="row.treeNode.children[0].row.entity.inventory_uuid"
```
Do you need this `id` field?  What is it used for?

lomamech replied:
Only during End to End test
````
  page.setHeaderValue = async function setHeaderValue(rowIndex, columnIndex, value) {
    const cell = await GU.getCell(gridId, rowIndex, columnIndex);
    const inputOutStock = cell.element(by.id('row.treeNode.children[0].row.entity.inventory_uuid'));
    inputOutStock.sendKeys(value);
  };
````

jniles commented on client/src/modules/stock/entry/modals/AggregateConsumptionModalForm.service.js:
```diff
@@ -0,0 +1,169 @@
+ * @description
+ * This function allows you to create a new instance to record
+ * the aggregate consumption of a batch by date. This function makes it possible to better track
+ * consumption to quickly detect periods of out of stock
```
Nice description!

jniles commented on client/src/modules/stock/aggregated_consumption/templates/days_stock_out.tmpl.html:
```diff
@@ -0,0 +1,12 @@
+<div class="ui-grid-cell-contents">
+  <input
+    ng-if="row.groupHeader"
+    name="days_stock_out"
```
Let's make this default to "0".

lomamech replied:
I wanted the user to write really zero instead of using a default value, this will prevent him from making the mistake of forgetting

jniles commented on server/controllers/stock/index.js:
```diff
@@ -975,3 +977,221 @@ function getStockTransfers(req, res, next) {
+            date : new Date(eltDate),
+            is_exit : 1,
+            flux_id : core.flux.TO_LOSS,
+            description : movement.description,
```
We should do what we do with the other stock movements and template in our own description as well.  Something like "Aggregated Stock Adjustment for Period X by User Y"

jniles commented on server/controllers/stock/reports/stock/exit_aggregate_consumption_receipt.js:
```diff
@@ -0,0 +1,82 @@
+  const report = new ReportManager(STOCK_AGGREGATE_CONSUMPTION_TEMPLATE, session, optionReport);
+
+  /**
+   * TODO consider the donor also in a donation transaction
```
Do you need this comment?  It looks irrelevant.

jniles commented on server/models/schema.sql:
```diff
@@ -834,8 +834,8 @@ CREATE TABLE `inventory_type` (
   `id` SMALLINT(5) UNSIGNED NOT NULL AUTO_INCREMENT,
-  `abbr` VARCHAR(10) NOT NULL,
-  `text` VARCHAR(30) NOT NULL,
+  `abbr` VARCHAR(100) NOT NULL,
```
👍

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.page.js:
```diff
@@ -0,0 +1,164 @@
+   */
+  page.setHeaderValue = async function setHeaderValue(rowIndex, columnIndex, value) {
+    const cell = await GU.getCell(gridId, rowIndex, columnIndex);
+    const inputOutStock = cell.element(by.id('row.treeNode.children[0].row.entity.inventory_uuid'));
```
This: `by.id('row.treeNode.children[0].row.entity.inventory_uuid')`  selector makes me think something is broken.  Shouldn't it be replaced by the inventory uuid?

I also suggest that you use the inventory text instead of the `id=inventory_uuid`.  Set something like `data-inventory-label="{{row.treeNode.children[0].row.entity.inventory_text}}" which will let you change this selector to:

```js
const inputOutStock = cell.$('[data-inventory-label="Amoxy ..."]')); 
```
which is much easier to read.

lomamech replied:
If I made this modification, I will also be forced to modify the server part to do the reverse operation to find the Uuid from the name of the inventory

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.page.js:
```diff
@@ -0,0 +1,164 @@
+    await getCell.element(by.id('aggregated_details')).click();
+  };
+
+  page.setLotsDetailled = async function setLotsDetailled(lotsArray) {
```
`detailed` only has a single `l` in English :)

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.page.js:
```diff
@@ -0,0 +1,164 @@
+    FU.modal.submit();
+  };
+
+  page.setLots2Detailled = async function setLots2Detailled(lotsArray) {
```
What is `setLots2`?  What is the actual action that is being performed here?

lomamech replied:
![image](https://user-images.githubusercontent.com/8035588/109496996-b296b580-7a91-11eb-8357-38232990fa8f.png)
It's just to fill the modal of consumption detailed by date during E2E test

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.page.js:
```diff
@@ -0,0 +1,164 @@
+    return components.depotSelect.set(depot, 'depot-supplier');
+  };
+
+  page.setRows = async number => {
```
Let's call this `addRows` instead of `setRows`, since we are clicking the "add rows" button on the page.

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.page.js:
```diff
@@ -0,0 +1,164 @@
+  page.modalGridId = modalGridId;
+
+  // create modal
+  page.showCreateModal = async isNewRequisition => {
```
Do we need the `isNewRequisition` flag here?

I think you may have copied this from somewhere.  Can you remove the methods you don't actually use from this page?

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.js:
```diff
@@ -0,0 +1,147 @@
+    await page.setFiscalPeriod(fiscalYearLabel, month[getMovementMonth]);
+    await page.setDescription(`Aggregate consumption from current depot ${DEPOT_TERTIAIRE}`);
+
+    await page.setHeaderValue(0, 7, 0);
```
In general, you should try not to refer to the index in the grid in your tests.  Instead, refer to the name of the row.  Something like
```js
await page.setHeaderValue("Quinine", 7);
```
or even better:
```js
const quinine = await page.getInventoryGroup('Quinine');
await quinine.setDaysOfStockOut(7);
await quinine.setLotQuantityLost("Lot X", 5);
await quinine.setLotQuantityConsumed("Lot X", 7);
```

This is much closer to what the user is actually doing.  It also means that other programmers can add lots/inventory items without needing to rewrite these tests.

jniles replied:
By the way, if you need help with this, I can assist.  I know my descriptions aren't always the best 👍

jniles commented on test/end-to-end/stock/stock.aggregate_consumption.js:
```diff
@@ -0,0 +1,147 @@
+
+    const getLastDays = new Date(getMovementYear, getMovementMonth + 1, 0);
+
+    const month = ['Jan', 'Fév', 'Mars', 'Avr', 'Mai', 'Juin', 'Juill', 'Août', 'Sept', 'Oct', 'Nov', 'Déc'];
```
You use these months a couple of times. Can you move them to be one of the first parts in the `describe` block?  That way, each `it()` will share them.

jniles commented on client/src/modules/stock/aggregated_consumption/modals/consumption_lot.modal.js:
```diff
@@ -0,0 +1,206 @@
+) {
+  const vm = this;
+
+  const cache = new AppCache('StockEntryModal');
```
I think this `AppCache` needs to be updated.

jniles commented on client/src/modules/stock/aggregated_consumption/templates/lot_movement_date.tmpl.html:
```diff
@@ -0,0 +1,7 @@
+<div class="ui-grid-cell-contents">
+  <bh-date-picker
```
I think this datepicker should have `min` and `max` dates within the period, right?

jniles commented on client/src/modules/stock/aggregated_consumption/modals/consumption_lot.modal.js:
```diff
@@ -0,0 +1,206 @@
+    footerCellClass : 'text-right',
+    cellTemplate : 'modules/stock/aggregated_consumption/templates/lost.tmpl.html',
+  }, {
+    field : 'actions',
```
It would be good to add `width: "10"` here, to limit the width of this field:

![image](https://user-images.githubusercontent.com/896472/109132837-8f4bcd80-7754-11eb-825c-a773675aa511.png)

jniles reviewed (commented):
@lomamech this represents a lot of work!  It's a very complicated module, and you've done an admirable job implementing it.  Well done!

A couple of improvements I would make.

If there are no days of stock out, I wouldn't offer the detailed consumption modal.  Specifically, I would hide these buttons:
![image](https://user-images.githubusercontent.com/896472/109128111-71c83500-774f-11eb-999b-56158cef5cf9.png)

It's true that maybe someone could make a RUMER-like consumption report, but realistically no one will do that.

Similar to the above comment, we also should hide the "details" link if both values are 0:
![image](https://user-images.githubusercontent.com/896472/109129354-db950e80-7750-11eb-9b0c-d2fce5cbbe29.png)

There appears to be two binding problems on the receipt modal and the receipt itself:
![image](https://user-images.githubusercontent.com/896472/109129048-79d4a480-7750-11eb-86c7-58825799ae5d.png)


Should we make the user choose a start date and end date on the `consumption_lot` modal instead of just a single date?  If the idea is that they can show multiple stock outs using it, then they should be able to show the start date, end date, and quantity consumed and lost over that period, no?

lomamech committed:
improvement(Aggregate consumption)
- Improved commit message
- Fixed the error on the migrate file
- Restrict the detailed consumption option only if the number of outage days is defined and if the quantity consumed or lost are also defined
- Correction of the display problem on the aggregated stock consumption receipt
- Improved error messages for invalid stock consumption
- Removed fast insert mode
- Use of a template for the description of aggregated stock consumption`

lomamech committed:
Alter table inventory_unit change tall of column

lomamech committed:
Use detailed instead detailled

lomamech committed:
Add property start date and End date during details of stock consumption

lomamech committed:
test(Aggregate consumption)
- Update and Complete some E2E test for aggregate consumption

lomamech committed:
Clean and sanitaze error

jniles commented on server/controllers/stock/index.js:
```diff
@@ -65,6 +65,10 @@ async function createStock(req, res, next) {
     const period = await Fiscal.lookupFiscalYearByDate(params.date);
     const periodId = period.id;
+
+    console.log('FISCAL_YEAR: _ID');
```
It looks like you forgot some console.logs() here.

Curious that hound didn't catch them.

lomamech replied:
It's Outdated

jniles commented on server/controllers/stock/index.js:
```diff
@@ -994,16 +998,28 @@ async function createAggregatedConsumption(req, res, next) {
 
     if (checkInvalid.length) {
-      throw new Error('Invalid data');
+      throw new Error('Invalid data!  Some lots have consumed or lost more stock than they originally had.');
```
Beautiful 👍

jniles reviewed (commented):
It looks like you did address the datepicker issue 👍 

Let me give it a test locally and see if the functionality works.  If so, I'll be happy to merge it.

I think the integration tests will fail, but we've put in fixes on the `refactor-stock-management` branch.  No need to work on them here.

jniles commented on client/src/modules/stock/aggregated_consumption/modals/consumption_lot.modal.js:
```diff
@@ -0,0 +1,212 @@
+    }
+  }
+
+  function getFirstEmptyLot() {
```
This is a great function name, by the way.  Super clear 👍

jniles commented on client/src/modules/stock/aggregated_consumption/modals/consumption_lot.modal.js:
```diff
@@ -0,0 +1,212 @@
+   * @param {string} rowLot the row.entity.lot string
+   */
+  function onLotBlur(rowLot) {
+    if (vm.enableFastInsert && rowLot) {
```
Do you ever use `enableFastInsert`?  I assume this code must be copied from somewhere.  We should probably remove anything that isn't critical for these modules.

This can be done after merge.

jniles commented on client/src/modules/stock/entry/modals/AggregateConsumptionModalForm.service.js:
```diff
@@ -0,0 +1,200 @@
+   * This function computes the total quantity consumed in the form.
+   */
+  StockForm.prototype.totalConsumed = function totalConsumed() {
+    return this.rows.reduce((sum, row) => sum + row.quantity_consumed, 0);
```
Cool! 👍

jniles commented on server/controllers/stock/reports/stock/exit/exitAggregateConsumption.js:
```diff
@@ -0,0 +1,39 @@
+
+/**
+ * @function fetch
+ * @description fetch stock exit for patient
```
I assume this description is out of date.  It should be updated.

jniles reviewed (approved):
@lomamech this is a much better step in the right direction.  I can see it coming together.

There are quite a few bugs that I discovered.  I don't think ti should block this - if we merge this into master, the whole team can dedicate their resources to helping fix them.

I'll list them here:

1. After setting the "details", you can update the consumed quantities.  I think the system will use the "detailed" version, and this can cause negative numbers!
2. It isn't entirely clear when the "details" link should appear.  This is partly my fault - I suggested you hide it.  I wonder if there is a way to disable it, and provide a tooltip to the user saying what needs to be done to enable it.
3. The validation of dates within the period work well, but it should limit the date picker to only those dates.  I screwed this up a couple of times.

Ultimately, this is a huge amount of work.  I'm really impressed and you should be proud of the work you've accomplished.

I'm going to manually begin the merging process.
